File: packages/typescript/spec/src/validation.ts
import type { ErrorObject } from "ajv";
import { Ajv2020 } from "ajv/dist/2020.js";
import addFormats = require("ajv-formats");

import canonicalSchema from "./generated/agent-topology.schema.json" with { type: "json" };
import type { ElementReference, TopologyDocument } from "./types.js";

export interface ValidationIssue {
  readonly path: string;
  readonly message: string;
}

export type ValidationResult =
  | { readonly valid: true; readonly document: TopologyDocument }
  | { readonly valid: false; readonly errors: readonly ValidationIssue[] };

const ajv = new Ajv2020({ allErrors: true, strict: true });
addFormats.default(ajv);
const validateSchema = ajv.compile<TopologyDocument>(canonicalSchema);

function cloneSchema(): object {
  return JSON.parse(JSON.stringify(canonicalSchema)) as object;
}

export function loadSchema(): object {
  return cloneSchema();
}

function schemaIssues(
  errors: ErrorObject[] | null | undefined,
): ValidationIssue[] {
  return (errors ?? [])
    .map((error) => ({
      path: error.instancePath === "" ? "$" : `$${error.instancePath}`,
      message: error.message ?? "is invalid",
    }))
    .sort(
      (left, right) =>
        left.path.localeCompare(right.path, "en") ||
        left.message.localeCompare(right.message, "en"),
    );
}

function duplicates(values: readonly string[]): string[] {
  const seen = new Set<string>();
  const duplicateValues = new Set<string>();
  for (const value of values) {
    if (seen.has(value)) duplicateValues.add(value);
    else seen.add(value);
  }
  return [...duplicateValues].sort();
}

function issue(path: string, message: string): ValidationIssue {
  return { path, message };
}

function referenceIssues(document: TopologyDocument): ValidationIssue[] {
  const errors: ValidationIssue[] = [];
  const graphIds = document.graphs.map((graph) => graph.id);
  for (const duplicate of duplicates(graphIds)) {
    errors.push(
      issue("$.graphs", `duplicate graph id ${JSON.stringify(duplicate)}`),
    );
  }

  const graphById = new Map(document.graphs.map((graph) => [graph.id, graph]));
  const elementIds = new Map<
    string,
    Record<ElementReference["kind"], Set<string>>
  >();

  document.graphs.forEach((graph, graphIndex) => {
    const structure = graph.structure;
    const structurePath = `$.graphs[${graphIndex}].structure`;
    const nodeIds = structure.nodes.map((node) => node.id);
    const edgeIds = structure.edges.map((edge) => edge.id);
    const joinIds = structure.joins.map((join) => join.id);
    const knownNodes = new Set(nodeIds);
    elementIds.set(graph.id, {
      graph: new Set([graph.id]),
      node: knownNodes,
      edge: new Set(edgeIds),
      join: new Set(joinIds),
    });

    for (const [kind, ids] of [
      ["node", nodeIds],
      ["edge", edgeIds],
      ["join", joinIds],
    ] as const) {
      for (const duplicate of duplicates(ids)) {
        errors.push(
          issue(
            `${structurePath}.${kind}s`,
            `duplicate ${kind} id ${JSON.stringify(duplicate)}`,
          ),
        );
      }
    }

    const requireNode = (reference: string, path: string): void => {
      if (!knownNodes.has(reference)) {
        errors.push(
          issue(path, `unknown node id ${JSON.stringify(reference)}`),
        );
      }
    };
    structure.nodes.forEach((node, nodeIndex) => {
      if (node.subgraphId !== undefined && !graphById.has(node.subgraphId)) {
        errors.push(
          issue(
            `${structurePath}.nodes[${nodeIndex}].subgraphId`,
            `unknown graph id ${JSON.stringify(node.subgraphId)}`,
          ),
        );
      }
    });
    structure.edges.forEach((edge, edgeIndex) => {
      requireNode(edge.source, `${structurePath}.edges[${edgeIndex}].source`);
      requireNode(edge.target, `${structurePath}.edges[${edgeIndex}].target`);
    });
    structure.joins.forEach((join, joinIndex) => {
      join.sources.forEach((source, sourceIndex) =>
        requireNode(
          source,
          `${structurePath}.joins[${joinIndex}].sources[${sourceIndex}]`,
        ),
      );
      requireNode(join.target, `${structurePath}.joins[${joinIndex}].target`);
    });
    for (const field of ["entryNodeIds", "exitNodeIds"] as const) {
      structure[field].forEach((reference, referenceIndex) =>
        requireNode(reference, `${structurePath}.${field}[${referenceIndex}]`),
      );
    }
  });

  const expectedStatus =
    document.completeness.gaps.length === 0 ? "complete" : "incomplete";
  if (document.completeness.status !== expectedStatus) {
    errors.push(
      issue(
        "$.completeness.status",
        `must be ${JSON.stringify(expectedStatus)} when gaps contains ${document.completeness.gaps.length} item(s)`,
      ),
    );
  }
  document.completeness.gaps.forEach((gap, gapIndex) => {
    const element = gap.element;
    const path = `$.completeness.gaps[${gapIndex}].element`;
    const ids = elementIds.get(element.graphId);
    if (ids === undefined) {
      errors.push(
        issue(
          `${path}.graphId`,
          `unknown graph id ${JSON.stringify(element.graphId)}`,
        ),
      );
    } else if (!ids[element.kind].has(element.id)) {
      errors.push(
        issue(
          `${path}.id`,
          `unknown ${element.kind} id ${JSON.stringify(element.id)} in graph ${JSON.stringify(element.graphId)}`,
        ),
      );
    }
  });
  return errors;
}

export function validateDocument(value: unknown): ValidationResult {
  if (!validateSchema(value)) {
    return { valid: false, errors: schemaIssues(validateSchema.errors) };
  }
  const document = value as TopologyDocument;
  const errors = referenceIssues(document);
  return errors.length === 0
    ? { valid: true, document }
    : { valid: false, errors };
}

export function isTopologyDocument(value: unknown): value is TopologyDocument {
  return validateDocument(value).valid;
}

export class TopologyValidationError extends Error {
  readonly errors: readonly ValidationIssue[];

  constructor(errors: readonly ValidationIssue[]) {
    super(`invalid agent-topology document (${errors.length} issue(s))`);
    this.name = "TopologyValidationError";
    this.errors = errors;
  }
}

export function assertTopologyDocument(
  value: unknown,
): asserts value is TopologyDocument {
  const result = validateDocument(value);
  if (!result.valid) throw new TopologyValidationError(result.errors);
}

File: packages/python/spec/src/agent_topology/spec/_validation.py
"""Validation against the packaged agent-topology document contract."""

from __future__ import annotations

import json
from collections.abc import Iterable
from importlib.resources import files
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator, FormatChecker


def load_schema() -> dict[str, Any]:
    """Load and return a fresh copy of the packaged canonical JSON Schema."""
    resource = files(__package__).joinpath("agent-topology.schema.json")
    if not resource.is_file():
        development_schema = next(
            (
                parent / "spec" / "agent-topology.schema.json"
                for parent in Path(__file__).resolve().parents
                if (parent / "spec" / "agent-topology.schema.json").is_file()
            ),
            None,
        )
        if development_schema is not None:
            return json.loads(development_schema.read_text(encoding="utf-8"))
    return json.loads(resource.read_text(encoding="utf-8"))


def _path(parts: Iterable[str | int]) -> str:
    result = "$"
    for part in parts:
        if isinstance(part, int):
            result += f"[{part}]"
        else:
            result += f".{part}"
    return result


def _duplicates(values: Iterable[str]) -> set[str]:
    seen: set[str] = set()
    duplicates: set[str] = set()
    for value in values:
        if value in seen:
            duplicates.add(value)
        else:
            seen.add(value)
    return duplicates


def _reference_errors(document: dict[str, Any]) -> list[str]:
    errors: list[str] = []
    graphs = document["graphs"]
    graph_ids = [graph["id"] for graph in graphs]

    for graph_id in sorted(_duplicates(graph_ids)):
        errors.append(f"$.graphs: duplicate graph id {graph_id!r}")

    graph_by_id = {graph["id"]: graph for graph in graphs}
    element_ids: dict[str, dict[str, set[str]]] = {}

    for graph_index, graph in enumerate(graphs):
        graph_path = f"$.graphs[{graph_index}]"
        structure_path = f"{graph_path}.structure"
        structure = graph["structure"]
        node_ids = [node["id"] for node in structure["nodes"]]
        edge_ids = [edge["id"] for edge in structure["edges"]]
        join_ids = [join["id"] for join in structure["joins"]]
        known_nodes = set(node_ids)
        element_ids[graph["id"]] = {
            "graph": {graph["id"]},
            "node": known_nodes,
            "edge": set(edge_ids),
            "join": set(join_ids),
        }

        for kind, ids in (("node", node_ids), ("edge", edge_ids), ("join", join_ids)):
            for duplicate in sorted(_duplicates(ids)):
                errors.append(
                    f"{structure_path}.{kind}s: duplicate {kind} id {duplicate!r}"
                )

        def require_node(reference: str, path: str) -> None:
            if reference not in known_nodes:
                errors.append(f"{path}: unknown node id {reference!r}")

        for node_index, node in enumerate(structure["nodes"]):
            subgraph_id = node.get("subgraphId")
            if subgraph_id is not None and subgraph_id not in graph_by_id:
                errors.append(
                    f"{structure_path}.nodes[{node_index}].subgraphId: "
                    f"unknown graph id {subgraph_id!r}"
                )

        for edge_index, edge in enumerate(structure["edges"]):
            require_node(edge["source"], f"{structure_path}.edges[{edge_index}].source")
            require_node(edge["target"], f"{structure_path}.edges[{edge_index}].target")

        for join_index, join in enumerate(structure["joins"]):
            for source_index, source in enumerate(join["sources"]):
                require_node(
                    source,
                    f"{structure_path}.joins[{join_index}].sources[{source_index}]",
                )
            require_node(join["target"], f"{structure_path}.joins[{join_index}].target")

        for field in ("entryNodeIds", "exitNodeIds"):
            for reference_index, reference in enumerate(structure[field]):
                require_node(reference, f"{structure_path}.{field}[{reference_index}]")

    gaps = document["completeness"]["gaps"]
    expected_status = "complete" if not gaps else "incomplete"
    actual_status = document["completeness"]["status"]
    if actual_status != expected_status:
        errors.append(
            "$.completeness.status: must be "
            f"{expected_status!r} when gaps contains {len(gaps)} item(s)"
        )

    for gap_index, gap in enumerate(gaps):
        element = gap["element"]
        path = f"$.completeness.gaps[{gap_index}].element"
        graph_id = element["graphId"]
        if graph_id not in element_ids:
            errors.append(f"{path}.graphId: unknown graph id {graph_id!r}")
            continue
        if element["id"] not in element_ids[graph_id][element["kind"]]:
            errors.append(
                f"{path}.id: unknown {element['kind']} id {element['id']!r} "
                f"in graph {graph_id!r}"
            )

    return errors


def validate_document(document: Any, schema: dict[str, Any] | None = None) -> list[str]:
    """Return deterministic, path-qualified errors for one topology document."""
    if schema is None:
        schema = load_schema()
    validator = Draft202012Validator(schema, format_checker=FormatChecker())
    schema_errors = sorted(
        validator.iter_errors(document),
        key=lambda error: (list(error.absolute_path), error.message),
    )
    errors = [
        f"{_path(error.absolute_path)}: {error.message}" for error in schema_errors
    ]
    if errors or not isinstance(document, dict):
        return errors
    return _reference_errors(document)
